README.md
# Mini-Calculator

## Introduction

This is a simple calculator built with the React library, using mathjs for the actual calculations. The logic observed here is **immediate execution**.

This calculator is project #4 of FreeCodeCamp(FCC)'s Front End Development Libraries Certification.

To see a detailed timeline of project progress, check out the changelog.

## Features

* input positive/negative integer/decimal numbers up to 22 digits.
* up to 22 digits in significant digits in terms of precision.
* scientific notation support.
* addition/subtraction/multiplication/division/modulo function.
* square/square root function.
* All Clear, Clear Entry, and backspace function.
* Up to 13 past calculations. (collapsible in mobile view)
* Set calculation to one of the past ones by clicking.
* numpad input support.


changelog.md
# v0.1 Brainstorm
## new
 * created app
 * installed gh-pages/mathjs

## to do
 * Basic arithmetic +/-/*//
 * Sqrt/^2
 * negative numbers
 * AC/Del
 * integer operations
 * decimal operations
 * process calculation on enter/on new operator
 * scientific notation toggle
 * ON/OFF switch
 
 * calculation history/undo
 * on the fly number edit
 * history change

# v0.3 Basics
## new
 * basic positive float calculation
 * clear all, clear entry, backspace function
 * removed scientific notation
 * no leading zero/dupe zeroes/dupe decimal
 * no dupe operators

### to do
 * sqrt/^2
 * negative numbers
 * consecutive calculations

 * calculation history
 * history change

# v0.4 Basics2
## new
 * sqrt/^2
 * negative numbers
 * consecutive calculations
 * scientific notationn
 * backspace function with scientific notation

### to do
 * refactor (extremely redundant and messy code)
 * FCC compliant negative number
 * FCC compliant display

 * calculation history
 * history change

# v0.5 MVP
## new
 * FCC compliant calculations
 * refactored (to best of ability)
 * calculation history
 * modulo function
 * fixed:
   * issue of state changes dependent on previous states, resulting in fast calculations generating unpredictable results
   * issue of negative sign being seen as a valid operand
   * issue of recording history reflecting inaccurate/incomplete calculations

### to do
 * SCSS
 * history change (time travel)
 * fix scientific notation not being properly counted in digit limit
 * truncation of numbers displayed in history
 * fix 30 digits too large to fit on display
 * add history size limit (20 perhaps?)

# v0.6 Feature update
## new
 * history change
 * scientific notation now properly added to digit limit
 * fixed sqrt/pow behavior on complete calculations
 * added dynamic font-size for extremely large numbers
 * added infinity for numbers with 10+ digits after e
 * added appropriate response to undefined/infinity as op2/res
 * added history number truncation
 * added history size limit

## in progress
 * collapsible history (mobile only, how to display only last item while collapsed, how to ensure all items fit when expanded)

### to do
 * SCSS
 * responsive design

# v0.7 almost release
## new
 * responsive design (react based) at 900px breakpoint
 * collapsible history at below 900px

## to do
 * source bg img
 * proper color scheme
 * test max history size and how it fits 

# v0.7 SCSS and troubleshooting
## new
 * fixed +/-, e+, pow, sqrt handling behavior to match expectations
 * removed console.logs
 * bg img
 * color scheme
 * history key hashing
 * operator behavior testing (cannot reproduce infinity multiplication bug, added additional check in calcing method as precaution)
 * reduce history size to 13
 * fixed power staying off when resizing viewport above breakpoint
 * split history display text to put result always on the right

## to do
 * split components
 * change readme
 * build&publish

# v1.0 build
## new
 * split components
 * update readme.md
 * deleted unnecessary comments
 * deleted redundant imgs
 * build

src/comp/Core.js
import React, {useState} from 'react';
import HistList from './HistList';
import Calc from './Calc';
import { create, all } from 'mathjs';


//dig. lim.
const digLim=22;
//sig. dig.
const config = {
  precision: digLim
};
const math = create(all, config);

//calculation core
//contains history & calc
const Core = props => {
    //left operand
    const [op1, setOp1] = useState('');
    //operator
    const [oper, setOper] = useState('');
    //right operand
    const [op2, setOp2] = useState('0');
    //if decimal 
    const [dec, setDec] = useState(false);
    //calculation result
    const [res, setRes] = useState('');
    //sci notation
    const [ey, setEy] = useState(false);
    //array of past history
    const [hist, setHist] = useState([]);
  
    //responsive (mobile)
    const mobSty={
      width:'400px',
      height:'100%',
      margin:'0 auto'
    }
  
    //responsive (desktop)
    const bigSty={
      width:'900px',
      height:'500px',
      display:'flex',
      top:`${Math.max((props.h-500)/2,0)}px`,
      left:`${Math.max((props.w-900)/2,0)}px`
    }
  
    //generate unique key based on time of action
    const getUnique=()=>{
      const d = new Date();
      return d.getMilliseconds().toString(16)+d.getDate().toString(16)+d.getMinutes().toString(16);
    }
    //record calculation to history when given all 4 parameters (max history size 13)
    const histing=(l,r,o,a)=>{
      const u=getUnique();
      if (hist.length>=13)
        setHist(hist.concat([[l,r,o,a,u]]).slice(1));
      else
      setHist(hist.concat([[l,r,o,a,u]]));
    }
    //calculate result given operands and operator
    const calcing = (l,r,o) => {
      if (l==='infinity' || r==='infinity')
        return 'infinity';
  
      //convert string to bignumber 
        const opL = math.bignumber(l);
        const opR = math.bignumber(r);
        let ans='';
        if (o === '+') {
          ans = (math.add(opL, opR).toString());
        }
        else if (o === '-') {
          ans = (math.subtract(opL, opR).toString());
        }
        else if (o === '*') {
          ans = (math.multiply(opL, opR).toString());
        }
        else if (o === '/') {
          //custom msg of dividing by zero
          if (r === '0')
            ans = ('undefined');
          else
            ans = (math.divide(opL, opR).toString());
        }
        else if (o ==='%') {
          ans = (math.mod(opL,opR).toString());
        }
  
        //if answer is too big (15 sig. dig. plus 'e+' plus decimal point leaves 10 digits for after e+)
        if (ans.length>28) 
          return 'infinity';
        return ans;
    }
  
    //distribute passed action to appropraite handler
    const handleEvent = e => {
      if (!props.power)
        return;
      if (e.length === 1) {
        if (e === '=') {
          handleEnt();
        }
        else if (e === 'e') {
          handleEy();
        }
        else if (/[\d.]/.test(e)) {
          handleNum(e);
        }
        else {
          handleArith(e);
        }
      }
      else if (e.length === 2) {
        handleClear(e);
      }
      else if (e === 'bksp') {
        handleDel();
      }
      else if (e === 'sign') {
        handleSign();
      }
      else
        handleOther(e);
    }
  
  
    //0-9 & decimal
    const handleNum = num => {
  
  
      //case of complete calculation, or undefined/infinity as part of operand
      //clear all, then process given input
      if (res !== '' || op2==='undefined' || op2==='infinity') {
        setOp1('');
        setOper('');
        setRes('');
        setEy(false);
        if (num === '.') {
          setOp2('0.');
          setDec(true);
        }
        else {
          setOp2(num);
          setDec(false);
        }
      }
      //max digit: 15
      else if (op2.length - (dec ? 1 : 0) - (ey ? 2:0) < digLim) {
        //dealing with leading zero
        if (num === '0') {
          //case of no number & case of decimal point present
          if (op2.length === 0 || dec)
            setOp2(op2.concat(num));
          //case of no already leading zero
          else if (op2[0] !== '0')
            setOp2(op2.concat(num));
        }
        //dealing with decimal
        else if (num === '.') {
          //case of no number
          if (op2.length === 0) {
            setDec(true);
            setOp2(op2.concat('0.'));
  
          }
          //case of no decimal and no scientific notation
          else if (!dec && !ey) {
            setDec(true);
            setOp2(op2.concat('.'));
  
          }
        }
        //dealing with 1-9
        else if (/[1-9]/.test(num)) {
          //case of only zero, no decimals
          if (op2.length > 0 && !dec && op2[0] === '0')
            setOp2(num);
          //all other cases
          else
            setOp2(op2.concat(num));
        }
  
  
      }
  
    }
  
    //+ - * / %
    const handleArith = op => {
      //case operand undefined/infinity or result undefined/infinity
      //clear all
      if (res==='undefined' || res==='infinity' || op2==='undefined' || op2==='infinity') {
        setOp1('');
        setOper('');
        setOp2('0');
        setRes('');
        setDec(false);
        setEy(false);
      }
      //case of complete calculation
      //half fill calculation with prev result as left operand, and given input as operator
      else if (res !== '') {
        const rem = res;
        setOp1(rem);
        setOper(op);
        setOp2('');
        setRes('');
        setDec(false);
        setEy(false);
      }
      //no operators
      else if (oper === '') {
        //case of existing operand (not just a negative sign)
        if (op2.length > 0 && op2[op2.length-1]!=='-') {
          //either no scientific notation, or a complete sci not (number must exist after e+)
          if (!ey || op2[op2.length - 1] !== '+') {
            const toMove = op2;
            setOp1(toMove);
            setOper(op);
            setOp2('');
            setRes('');
            setDec(false);
            setEy(false);
          }
  
        }
      }
      //operator already exists
      else {
        //case of no existing op2, or op2 is just a negative sign
        if (op2.length === 0||op2[op2.length-1]==='-') {
          //case of - operator, swap pos/neg
          if (op==='-') {
            handleSign();
          }
          //case of other operators (clear possible lone negative sign)
          else {
            setOper(op);
            setOp2('');
          }
  
        }
        //case of two operands
        else {
          //either no scientific notation, or a complete sci not (number must exist after e+) and its not just a - sign
          if ((!ey || op2[op2.length - 1] !== '+') && op2[op2.length - 1] !== '-') {
  
            const oOld=oper;
            const l=op1;
            const r=op2;
            const rekt = calcing(l,r,oOld);
  
            setOp1(rekt);
            setOper(op);
            setOp2('');
            setRes('');
            setDec(false);
            setEy(false);
            histing(l,oOld,r,rekt);
          }
  
        }
      }
  
    }
  
    // enter (equals)
    const handleEnt = () => {
      //only works with existing op1, op2, and a operator, with no completed calculation in place
      if (res === '' && op1.length > 0 && oper !== '' && op2.length > 0) {
        //additional requirements of:
        //either no scientific notation, or a complete sci not (number must exist after e+) 
        //not just a - sign as entire operand
        if ((!ey || op2[op2.length - 1] !== '+') && op2[op2.length - 1] !== '-') {
          const rekt=calcing(op1,op2,oper);
          setRes(rekt);
          histing(op1,oper,op2,rekt);
        }
      }
    }
  
    //handle +/- (test condition to see if behaves in expected way)
    const handleSign = () => {
      //case of complete calculation or operand is undefined/infinity
      if (res !== '' || op2==='undefined' || op2==='infinity') {
        setOp1('');
        setOper('');
        setOp2('-');
        setRes('');
        setDec(false);
        setEy(false);
      }
      //case of existing op2
      else if (op2.length > 0) {
        //case of positive to negative
        if (op2[0] !== '-')
          setOp2('-'.concat(op2));
        //case of negative to positive
        else
          setOp2(op2.slice(1));
      }
      //case of empty right operand
      else
        setOp2('-');
    }
  
    //handle sci notation (test removal feature)
    const handleEy = () => {
      //only works with existing op2 that is not undefined/infinity
      //only works with incomplete calculation
      if (res==='' && op2.length > 0 && op2!=='undefined' && op2!=='infinity') {
        //remove sci note. if already exists
        if (ey) {
          const newOp2=op2.slice(0,op2.indexOf('e'));
          setOp2(newOp2);
          setEy(false);
        }
        //not only 0 as operand, not trailing decimal place, sci notation doesn't already exist
        else if (op2[0] !== '0' || (dec && op2.length > 2) || !ey) {
          setOp2(op2.concat('e+'));
          setEy(true);
        }
      }
    }
    //sqrt, squared
    const handleOther = op => {
      //case of complete calculation and res not infinity/undefined
      if (res!=='' && res!=='undefined' && res!=='infinity') {
        let rem=math.bignumber(res);
        if (op === 'sqrt') {
          rem=math.sqrt(rem).toString();
        }
        else {
          rem=math.pow(rem,2).toString();
        }
        //perform operation based on input, then clear all and set op2
        setOp1('');
        setOper('');
        setOp2(rem);
        setRes('');
        setDec(rem.includes('.'));
        setEy(rem.includes('e'));
      }
      //only works with existing op2 and incomplete calculations
      else if (op2.length > 0 && (op2!=='undefined'&&op2!=='infinity')) {
        const toCalc = math.bignumber(op2);
        if (op === 'sqrt') {
          const ans = math.sqrt(toCalc).toString();
          setOp2(ans.length>28?'infinity':ans);
          //set sci notation if exists
          setEy(ans.includes('e'));
        }
        else {
          const ans = math.pow(toCalc, 2).toString();
          setOp2(ans.length>28?'infinity':ans);
          //set sci notation if exists
          setEy(ans.includes('e'));
        }
      }
  
    }
  
    //clears
    const handleClear = e => {
      //all clear
      if (e === 'ac') {
        setOp1('');
        setOper('');
        setOp2('0');
        setRes('');
        setDec(false);
        setEy(false);
      }
      //clear entry
      else {
        //case of calculation completed, or both operands exist
        //remove both result and right operand
        if (op1.length > 0 && op2.length > 0) {
          const saveOp1 = op1;
          const saveOp = oper;
          setOp1(saveOp1);
          setOper(saveOp);
          setOp2('');
          setRes('');
          setDec(false);
          setEy(false);
        }
        //case of 1 operand plus operator
        //remove operator, move left operand to right operand
        else if (op1.length > 0 && oper) {
          const saveOp1 = op1;
          setOp1('');
          setOper('');
          setOp2(saveOp1);
          setRes('');
          setEy((/e\+/i).test(saveOp1));
          setDec((/\./).test(saveOp1));
        }
        //case of only 1 operand
        //clear existing right operand
        else {
          setOp2('');
        }
      }
    }
  
    //time travel given index
    const handleHist = i => {
      const toSet=hist[i];
      setOp1(toSet[0]);
      setOper(toSet[1]);
      setOp2(toSet[2]);
      setRes(toSet[3]);
    }
  
    //handle backspaces
    const handleDel = () => {
      //no complete calculation only
      if (res === '') {
        //case existing right operand
        if (op2.length > 0) {
          //case of on the scientific notation 
          if (op2[op2.length - 1] === '+') {
            setOp2(op2.slice(0, op2.length - 2));
            setEy(false);
          }
          //case of on the decimal
          else if (op2[op2.length - 1] === '.') {
            setOp2(op2.slice(0, op2.length - 1));
            setDec(false);
          }
          //case of 0-9 or -
          else {
            setOp2(op2.slice(0, op2.length - 1));
          }
        }
        //case left operand and operator
        //remove operator and move left operand to right operand
        else if (op1.length > 0 && oper) {
          const saveOp1 = op1;
          setOp1('');
          setOper('');
          setOp2(saveOp1);
          setRes('');
          setEy((/e\+/i).test(saveOp1));
          setDec((/\./).test(saveOp1));
        }
      }
    }
  
    return (
      <div id="core"
        style={props.w>900?{...bigSty}:{...mobSty}}>
        <HistList hist={hist} 
          h={props.h}
          w={props.w}
          handleHist={handleHist}
          handlePow={props.handlePow}
          power={props.power}/>
        <Calc 
          h={props.h}
          w={props.w}
          op1={op1}
          oper={oper}
          op2={op2}
          res={res}
          hist={hist}
          handleEvent={handleEvent} />
      </div>
    )
  }

  export default Core;